pkg/ingest/store/local.go

```diff
@@ -0,0 +1,165 @@
+package store
+
+import (
+	"context"
+	"crypto/md5" //nolint:gosec
+	"encoding/hex"
+	"encoding/json"
+	"io"
+	"io/fs"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	nanoid "github.com/matoous/go-nanoid/v2"
+	"golang.org/x/exp/slices"
+)
+
+const cacheDirName = "_lakefs_cache"
+
+type LocalWalker struct {
+	mark            Mark
+	allowedPrefixes []string
+	cacheLocation   string
+}
+
+func NewLocalWalker(blockStorePath string, allowedPrefixes []string) *LocalWalker {
+	var cacheLocation string
+	if blockStorePath != "" {
+		cacheLocation = filepath.Join(blockStorePath, cacheDirName)
+	}
+	return &LocalWalker{
+		allowedPrefixes: allowedPrefixes,
+		mark:            Mark{HasMore: true},
+		cacheLocation:   cacheLocation,
+	}
+}
+
+func (l *LocalWalker) Walk(_ context.Context, storageURI *url.URL, options WalkOptions, walkFn func(e ObjectStoreEntry) error) error {
+	if storageURI.Scheme != "local" {
+		return path.ErrBadPattern
+	}
+	root := path.Join(storageURI.Host, storageURI.Path)
+	if err := l.verifyAbsPath(root); err != nil {
+		return err
+	}
+
+	var entries []*ObjectStoreEntry
+	// verify and use cache - location is stored in continuation token
+	if options.ContinuationToken != "" && strings.HasPrefix(options.ContinuationToken, l.cacheLocation) {
+		cacheData, err := os.ReadFile(options.ContinuationToken)
+		if err == nil {
+			err = json.Unmarshal(cacheData, &entries)
+			if err != nil {
+				entries = nil
+			} else {
+				l.mark.ContinuationToken = options.ContinuationToken
+			}
+		}
+	}
+
+	// if needed scan all entries to import and calc etag
+	if entries == nil {
+		if err := filepath.Walk(root, func(p string, info fs.FileInfo, err error) error {
```

Comment:
Can we at least use [filepath.WalkDir](https://pkg.go.dev/path/filepath#WalkDir) instead?  It is a bit less inefficient.

Reply:
Ouch.  My ideal implementation would support `d_type` on Linux, avoiding the extra stat call.  I remember this used to be a [point of contention](https://bugzilla.redhat.com/show_bug.cgi?id=1248140); I don't know where things ended up on the Linux NFS drivers, especially whether they can get the field from cooperative NASes.

This is not academic: importing will hammer the NAS if we stat every file.